javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java

```diff
@@ -0,0 +1,57 @@
+package com.semmle.js.dependencies;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class VersionConstraint {
+    private String operator;
+    private SemVer version;
+
+    public VersionConstraint(String operator, SemVer version) {
+        this.operator = operator;
+        this.version = version;
+    }
+
+    public static VersionConstraint exact(String version) {
+        return new VersionConstraint("", SemVer.tryParse(version));
+    }
+
+    public String getOperator() {
+        return operator;
+    }
+
+    public SemVer getVersion() {
+        return version;
+    }
+
+    private static final Pattern pattern = Pattern.compile("([~^<>=]*)\\s*(\\d.*)");
+
+    public static List<VersionConstraint> parseVersionConstraints(String str) {
+        String[] parts = str.split(",");
+        List<VersionConstraint> constraints = new ArrayList<>();
+        for (String part : parts) {
+            part = part.trim();
+            if (part.equals("*")) {
+                constraints.add(new VersionConstraint("*", null));
+                continue;
+            }
+            Matcher matcher = pattern.matcher(str);
+            if (matcher.matches()) {
+                String operator = matcher.group(1);
+                String versionStr = matcher.group(2);
+                if (operator.isEmpty() && versionStr.contains("x")) {
+                    // Normalize "1.x" to ">= 1.0"
```

Comment:
How about `1.0.x`? 

Also: "1.x" and ">= 1.0" are not equivalent (https://semver.npmjs.com/).

Reply:
Right, I'll make it a little more clear that we just parse the constraints in a way that works for our resolution algorithm (i.e. when picking only explicitly-mentioned versions, the range semantics aren't important). We just treat the `x` as a zero for this purpose.